### ISSUE TITLE
Regex optimisation on index does redundant lookups

### ISSUE THREAD

**Issue by gouthamve** [cortexproject/cortex#2906]
When caching index entries, we basically lookup the entire row (to cache it) and then filter the entries that match after loading the full row: https://github.com/cortexproject/cortex/blob/071502d118a1d0687ef21f832b196341e20bdad1/pkg/chunk/util/util.go#L84-L116 and https://github.com/cortexproject/cortex/blob/071502d118a1d0687ef21f832b196341e20bdad1/pkg/chunk/storage/caching_index_client.go#L70-L71

Now we introduced an optimisation that does the following: https://github.com/cortexproject/cortex/blob/071502d118a1d0687ef21f832b196341e20bdad1/pkg/chunk/chunk_store.go#L463-L470

Basically if we have a lookup that looks like `label=~"a|b|c|d"` we are now splitting it into 4 different lookups of `label=a`, `label=b`, etc. This is causing us to load the entire row multiple times due to caching. We should make sure to push this optmisation down to `filteringBatchIter` and not load something multiple times.
